mpsl: cx: Matched nrf7002 with nrf700x coex implementation (#8750)
nrfconnect/sdk-nrf: doc/nrf/ug_radio_coex.rst

.. figure:: images/coex_generic_3pin_pta.svg
   :alt: PTA interface supported by the ``nrf_radio_coex`` node with an nRF52 Series SoC

.. _ug_radio_cx_nrf700x_hw_if:

nRF700x Coexistence hardware interface
======================================

To configure the hardware for a PTA inside the nRF700x series chips:

1. Add the following node in the devicetree source file:

   .. code-block::

      / {
            nrf_radio_coex: nrf7002-coex {
               status = "okay";
               compatible = "nordic,nrf700x-coex";
               req-gpios =     <&gpio0 24 (GPIO_ACTIVE_HIGH)>;
               pri-dir-gpios = <&gpio0 14 (GPIO_ACTIVE_HIGH)>;
               grant-gpios =   <&gpio0 25 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
         };
      };

#. Optionally replace the node name ``nrf7002-coex`` with a custom one.
#. Replace the pin numbers provided for each of the required properties:

   * ``req-gpios`` - GPIO characteristic of the device that controls the ``REQUEST`` signal of the PTA.
   * ``pri-dir-gpios`` - GPIO characteristic of the device that controls the ``PRIORITY`` signal of the PTA.

>    * ``grant-gpios`` - GPIO characteristic of the device that controls the ``GRANT`` signal of the PTA (RF medium access granted).

Reviewer: When this comment is addressed https://github.com/nrfconnect/sdk-nrf/pull/8750/files#r972850052
there will probably a need to rename pins (`...that controls the PRIORITY`)

Author: Fixed

The following CX implementations are available:

* :ref:`ug_radio_cx_nrf700x_impl`
* :ref:`ug_radio_cx_thread_impl`

Each implementation is a plugin composed of one or more C source files interfacing with the MPSL API, which, in turn, communicates with the SR protocol drivers.

When one of the CX variants is enabled, the driver requests access to the RF medium from the PTA and informs it about the operation it is about to perform.
It also reacts properly to the information from the PTA that access to the RF medium was either granted or denied.

.. _ug_radio_cx_nrf700x_impl:

nRF700x interface implementation

> ================================

Reviewer: Again, if https://github.com/nrfconnect/sdk-nrf/pull/8750/files#r972850052 is addressed the pin names
below (REQUEST, PRIORITY, GRANT) will possible need to be renamed.

  It contains information about the type of operation (either RX or TX) to perform.
  It can be configured by setting the ``status0-gpios`` property of the ``nrf_radio_coex`` devicetree node.
* GRANT pin - It is the input of the SoC controlled by the PTA.
  It asserts when PTA grants access to the RF to the 802.15.4 and deasserts when it denies the access.
  It can be configured by setting the ``grant-gpios`` property of the ``nrf_radio_coex`` devicetree node.

The support for this interface is provided in a single-file plugin located in the sdk-nrf repo, in the :file:`subsys/mpsl/cx` directory.

Adding support for the CX nRF700x interface
-------------------------------------------

The nRF700x interface is supported out of the box.
To use it, complete the following steps:

1. Add the devicetree `nrf_radio_coex` node as described in :ref:`ug_radio_cx_nrf700x_hw_if`.
2. If :kconfig:option:`CONFIG_WIFI` is set to ``y`` and one of the supported SR protocols is enabled no additional steps are needed.
   If you do not wish to use the radio coexistence implementation, you must set  :kconfig:option:`CONFIG_MPSL_CX` to ``n``

> 3. If :kconfig:option:`CONFIG_WIFI` is set to ``n`` or none of the supported SR protocols is enabled, the implementation must be enabled manually.
>    This is done by setting both :kconfig:option:`CONFIG_MPSL_CX` and :kconfig:option:`CONFIG_MPSL_CX_NRF700X ` to ``y``.

Reviewer 2: ```suggestion
3. If :kconfig:option:`CONFIG_WIFI` is disabled or none of the supported SR protocols are enabled the implementation must be enabled manually with :kconfig:option:`CONFIG_MPSL_CX` and :kconfig:option:`CONFIG_MPSL_CX_NRF700X `.
```

* PRIORITY pin - It is the output controlled by the protocol driver.
  It contains information about the type of operation (either RX or TX) to perform.
  It can be configured by setting the ``pri-dir-gpios`` property of the ``nrf_radio_coex`` devicetree node.

> * GRANT pin - It is the input of the SoC controlled:kconfig:option:`CONFIG_MPSL_CX` by the PTA.

Reviewer 3: This change seems to be unwanted (looks like an accidental insertion).

Author: fixed

> * GRANT pin - It is the input of the SoC controlled:kconfig:option:`CONFIG_MPSL_CX` by the PTA.

Reviewer: seems this is improperly injected
```
`:kconfig:option:`CONFIG_MPSL_CX`
```

Author: Fixed
